### PR TITLE
Update DCGM receiver to detect supported profiling fields on a given GPU model

### PR DESCRIPTION
**Pull request by LujieDuan**
Overview: 
- Modify the DCGM receiver to support GPUs that do not emit all profiling metrics the receiver wants.

Implementation:
- On the start up of the receiver, calling `dcgm.GetSupportedMetricGroups()` to get all the available profiling metrics of the current GPUs, and use that decide which fields to set watch group on.

How is this tested:
- Modify the unit tests (the `has_gpu` tests) to use golden files. The golden files can be generated for each GPU model by: `go test -tags=gpu,has_gpu  -race ./receiver/dcgmreceiver/ -update`.

### CODE REVIEW

**Review by braydonk — COMMENTED** [review]
Review is just about Go code stuff, don't have any context on the changes themselves.

File: receiver/dcgmreceiver/client.go [hunk 1]
```diff
@@ -180,6 +190,63 @@ func discoverEnabledFieldIDs(config *Config) []dcgm.Short {
 	return enabledFieldIDs
 }
 
+// getAllSupportedFields calls the DCGM query function to find out all the
+// fields that are supported by the current GPUs
+func getAllSupportedFields() ([]dcgm.Short, error) {
+	// Fileds like `DCGM_FI_DEV_*` are not profiling fields, and they are alwasy
+	// supported on all devices
+	var supported = []dcgm.Short{
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client.go]: Updated, thanks!

File: receiver/dcgmreceiver/client.go [hunk 2]
```diff
@@ -180,6 +190,63 @@ func discoverEnabledFieldIDs(config *Config) []dcgm.Short {
...
+		dcgm.DCGM_FI["DCGM_FI_DEV_FB_USED"],
+		dcgm.DCGM_FI["DCGM_FI_DEV_FB_FREE"],
+	}
+	// GetSupportedMetricGroups currently does not support passing the actual
+	// group handle; here we pass 0 to query supported fields for group 0, which
+	// is the default DCGM group that is **supposed** to include all GPUs of the
+	// host.
+	fieldGroups, err := dcgm.GetSupportedMetricGroups(0)
+	if err != nil {
+		if dcgmErr, ok := err.(*dcgm.DcgmError); ok {
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client.go]: Done!

File: receiver/dcgmreceiver/client_gpu_test.go [hunk 3]
```diff
@@ -21,49 +21,135 @@ package dcgmreceiver
...
+		t.Fatal(err)
+	}
+	assert.Equal(t, len(dcgmNameToMetricName), len(client.enabledFieldIDs)+len(offFields))
+	goldenPath := getModelGoldenFilePath(t, gpuModel)
+	golden.Assert(t, string(actual), goldenPath)
+}
+
+// LoadExpectedMetrics read the supported metrics of a GPU model from the golden
+// file, given a GPU model string
+func LoadExpectedMetrics(t *testing.T, model string) []string {
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client_gpu_test.go:93]: This function is also used in the scraper test, so will need to export this. But I realized I do not need to export the struct `ModelSupportedFields`, so fixed that. Thanks for the comment!

File: receiver/dcgmreceiver/client_gpu_test.go [hunk 4]
```diff
@@ -21,49 +21,135 @@ package dcgmreceiver
...
+	var expectedMetrics []string
+	for _, supported := range m.SupportedFields {
+		expectedMetrics = append(expectedMetrics, dcgmNameToMetricNameMap[supported])
+	}
+	return expectedMetrics
+}
+
+// getModelGoldenFilePath returns golden file path given a GPU model string
+func getModelGoldenFilePath(t *testing.T, model string) string {
+	testDir, err := os.Getwd()
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client_gpu_test.go:131]: Done!

**Review by braydonk — APPROVED** [review]
All good from my side, still needs approval from someone with GPU context.

(Optional: I realized after looking again that there's a few other spots `t.Helper` may be desirable, I'll leave it up to you though)

File: receiver/dcgmreceiver/client_gpu_test.go [hunk 5]
```diff
@@ -21,49 +21,137 @@ package dcgmreceiver
...
+	Model string `yaml:"model"`
+	// List of supported fields
+	SupportedFields []string `yaml:"supported_fields"`
+	// List of unsupported fields
+	UnsupportedFields []string `yaml:"unsupported_fields"`
+}
+
+// TestSupportedFieldsWithGolden test getAllSupportedFields() against the golden
+// files for the current GPU model
+func TestSupportedFieldsWithGolden(t *testing.T) {
```
Comment (suffiank) [inline receiver/dcgmreceiver/client_gpu_test.go:53]: Nice! :)

**Review by suffiank — COMMENTED** [review]
LGTM ✅ I like the way you adjusted to tests to account for GPU models. Just some minor comments.

**Review by suffiank — APPROVED** [review]

File: receiver/dcgmreceiver/client.go [hunk 6]
```diff
@@ -70,15 +71,25 @@ func newClient(config *Config, logger *zap.Logger) (*dcgmClient, error) {
 	}
 
 	enabledFieldIDs := discoverEnabledFieldIDs(config)
-	enabledFieldGroup, err := setWatchesOnEnabledFields(config, logger, deviceGroup, enabledFieldIDs)
+	supportedFieldIDs, err := getAllSupportedFields()
+	if err != nil {
+		// If there is error querying the supported fields at all, let the
+		// receiver collect basic metrics: (GPU utilization, used/free memory).
+		logger.Sugar().Warnf("error querying supported profiling fields; only basic metrics will be collected: %w", err)
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client.go]: Done! Thanks!

File: receiver/dcgmreceiver/client.go [hunk 7]
```diff
@@ -180,6 +191,64 @@ func discoverEnabledFieldIDs(config *Config) []dcgm.Short {
...
+// fields that are supported by the current GPUs
+func getAllSupportedFields() ([]dcgm.Short, error) {
+	// Fields like `DCGM_FI_DEV_*` are not profiling fields, and they are always
+	// supported on all devices
+	supported := []dcgm.Short{
+		dcgm.DCGM_FI["DCGM_FI_DEV_GPU_UTIL"],
+		dcgm.DCGM_FI["DCGM_FI_DEV_FB_USED"],
+		dcgm.DCGM_FI["DCGM_FI_DEV_FB_FREE"],
+	}
+	// GetSupportedMetricGroups currently does not support passing the actual
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client.go:204]: Good idea - opened an issue https://github.com/NVIDIA/go-dcgm/issues/32 to check.

File: receiver/dcgmreceiver/client.go [hunk 8]
```diff
@@ -70,15 +71,25 @@ func newClient(config *Config, logger *zap.Logger) (*dcgmClient, error) {
...
 
 	enabledFieldIDs := discoverEnabledFieldIDs(config)
-	enabledFieldGroup, err := setWatchesOnEnabledFields(config, logger, deviceGroup, enabledFieldIDs)
+	supportedFieldIDs, err := getAllSupportedFields()
+	if err != nil {
+		// If there is error querying the supported fields at all, let the
+		// receiver collect basic metrics: (GPU utilization, used/free memory).
+		logger.Sugar().Warnf("error querying supported profiling fields; only basic metrics will be collected: %w", err)
+	}
+	onFields, offFields := filterSupportedFields(enabledFieldIDs, supportedFieldIDs)
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply receiver/dcgmreceiver/client.go]: Done! Those are much better naming :)

File: receiver/dcgmreceiver/client_gpu_test.go (hunk 5 above)
Comment (LujieDuan) [reply receiver/dcgmreceiver/client_gpu_test.go:53]: :) Thanks!

**Pull request merged**